src/sinks/util/service/health.rs

```diff
@@ -0,0 +1,324 @@
+}
+
+impl HealthConfig {
+    pub fn build<S, L>(
+        &self,
+        logic: L,
+        inner: S,
+        open: OpenGauge,
+        endpoint: String,
+    ) -> HealthService<S, L> {
+        let counters = Arc::new(HealthCounters::new());
+        let snapshot = counters.snapshot();
+
+        HealthService {
+            inner,
+            logic,
+            counters,
+            snapshot,
+            endpoint,
+            state: CircuitState::Closed(open.clone().open(emit_active_endpoints)),
+            open,
+            // An exponential backoff starting from retry_initial_backoff_sec and doubling every time
+            // up to retry_max_duration_secs.
+            backoff: ExponentialBackoff::from_millis(2)
+                .factor(
+                    (self
+                        .retry_initial_backoff_secs
+                        .unwrap_or(RETRY_INITIAL_BACKOFF_SECONDS_DEFAULT)
+                        .saturating_mul(1000)
+                        / 2)
+                    .max(1),
+                )
+                .max_delay(Duration::from_secs(
+                    self.retry_max_duration_secs
+                        .unwrap_or(RETRY_MAX_DURATION_SECONDS_DEFAULT),
+                )),
+        }
+    }
+}
+
+pub trait HealthLogic: Clone + Send + Sync + 'static {
+    type Error: Send + Sync + 'static;
+    type Response;
+
+    /// Returns health of the endpoint based on the response/error.
+    /// None if there is not enough information to determine it.
+    fn is_healthy(&self, response: &Result<Self::Response, Self::Error>) -> Option<bool>;
+}
+
+enum CircuitState {
+    /// Service is unhealthy hence it's not passing requests downstream.
+    /// Contains timeout.
+    Open(BoxFuture<'static, ()>),
+
+    /// Service will pass one request to test its health.
+    HalfOpen {
+        permit: Option<Trigger>,
+        done: Tripwire,
+    },
+
+    /// Service is healthy and passing requests downstream.
+    Closed(OpenToken<fn(usize)>),
+}
+
+/// A service which monitors the health of a service.
+/// Behaves like a circuit breaker.
+pub struct HealthService<S, L> {
+    inner: S,
+    logic: L,
+    counters: Arc<HealthCounters>,
+    snapshot: HealthSnapshot,
+    backoff: ExponentialBackoff,
+    state: CircuitState,
+    open: OpenGauge,
+    endpoint: String,
+}
+
+impl<S, L, Req> Service<Req> for HealthService<S, L>
+where
+    L: HealthLogic<Response = S::Response, Error = S::Error>,
+    S: Service<Req>,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = HealthFuture<S::Future, L>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        loop {
+            self.state = match self.state {
+                CircuitState::Open(ref mut timer) => {
+                    ready!(timer.as_mut().poll(cx));
+
+                    debug!(message="Endpoint is on probation.", endpoint = %&self.endpoint);
```

Comment:
```suggestion
                    debug!("Endpoint is on probation.", endpoint = %&self.endpoint);
```

Reply:
Compiler complains without `message=` since then it tries to parse `endpoint = %&self.endpoint` as an expression.

Reply:
Ah, fair enough - then just minor nit that there aren't spaces around the `=` :smile:

Reply:
(which I see you've done)